<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.13"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>GeneticAlgorithms: GeneticAlgorithms.py File Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">GeneticAlgorithms
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.13 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle">
<div class="title">GeneticAlgorithms.py File Reference</div>  </div>
</div><!--header-->
<div class="contents">
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGeneticAlgorithms_1_1__ThreadWithReturnValue.html">GeneticAlgorithms._ThreadWithReturnValue</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Class derived from Thread that returns the return value of the threaded target function in the return value of <a class="el" href="classGeneticAlgorithms_1_1__ThreadWithReturnValue.html#af628ff062d0ede3a96ebccd1ec66e192" title="Overridden join method returning return value. ">join()</a>.  <a href="classGeneticAlgorithms_1_1__ThreadWithReturnValue.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGeneticAlgorithms_1_1Genotype.html">GeneticAlgorithms.Genotype</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Object representation of a genotype including genome and its properties and fitness of the corresponding phenotype in a given environment provided by the objective function.  <a href="classGeneticAlgorithms_1_1Genotype.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGeneticAlgorithms_1_1GeneticAlgorithms.html">GeneticAlgorithms.GeneticAlgorithms</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Genetic Algorithms (GA) optimizer for arbitrary objective functions.  <a href="classGeneticAlgorithms_1_1GeneticAlgorithms.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGeneticAlgorithms_1_1TestGenericAlgorithms.html">GeneticAlgorithms.TestGenericAlgorithms</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Unit Test Class - all methods starting with test are automatically executed by unittest.  <a href="classGeneticAlgorithms_1_1TestGenericAlgorithms.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:ad4e5726cd3b24d9b11968b7a920c28c5"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="GeneticAlgorithms_8py.html#ad4e5726cd3b24d9b11968b7a920c28c5">GeneticAlgorithms.printHelp</a> ()</td></tr>
<tr class="memdesc:ad4e5726cd3b24d9b11968b7a920c28c5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print command line help text.  <a href="GeneticAlgorithms_8py.html#ad4e5726cd3b24d9b11968b7a920c28c5">More...</a><br /></td></tr>
<tr class="separator:ad4e5726cd3b24d9b11968b7a920c28c5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa9c763ce6c1837e80b6dc3abb2bbb2b5"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="GeneticAlgorithms_8py.html#aa9c763ce6c1837e80b6dc3abb2bbb2b5">GeneticAlgorithms.objfunc0</a> (args)</td></tr>
<tr class="memdesc:aa9c763ce6c1837e80b6dc3abb2bbb2b5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Objective function for demo from Goldberg's book (modified).  <a href="GeneticAlgorithms_8py.html#aa9c763ce6c1837e80b6dc3abb2bbb2b5">More...</a><br /></td></tr>
<tr class="separator:aa9c763ce6c1837e80b6dc3abb2bbb2b5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad568d0ddc86966d7809d304b52765e88"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="GeneticAlgorithms_8py.html#ad568d0ddc86966d7809d304b52765e88">GeneticAlgorithms.objfunc1</a> (args)</td></tr>
<tr class="memdesc:ad568d0ddc86966d7809d304b52765e88"><td class="mdescLeft">&#160;</td><td class="mdescRight">Objective function 1 for demo.  <a href="GeneticAlgorithms_8py.html#ad568d0ddc86966d7809d304b52765e88">More...</a><br /></td></tr>
<tr class="separator:ad568d0ddc86966d7809d304b52765e88"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a84f4a4e55c489adb038b679b81e0beae"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="GeneticAlgorithms_8py.html#a84f4a4e55c489adb038b679b81e0beae">GeneticAlgorithms.objfunc2</a> (args)</td></tr>
<tr class="memdesc:a84f4a4e55c489adb038b679b81e0beae"><td class="mdescLeft">&#160;</td><td class="mdescRight">Objective function 2 for demo.  <a href="GeneticAlgorithms_8py.html#a84f4a4e55c489adb038b679b81e0beae">More...</a><br /></td></tr>
<tr class="separator:a84f4a4e55c489adb038b679b81e0beae"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9b341db0c7dda40ecd30bc6491a0832f"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="GeneticAlgorithms_8py.html#a9b341db0c7dda40ecd30bc6491a0832f">GeneticAlgorithms.tsp</a> (args)</td></tr>
<tr class="memdesc:a9b341db0c7dda40ecd30bc6491a0832f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Traveling Salesman Problem objective function.  <a href="GeneticAlgorithms_8py.html#a9b341db0c7dda40ecd30bc6491a0832f">More...</a><br /></td></tr>
<tr class="separator:a9b341db0c7dda40ecd30bc6491a0832f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a219a0bde9e1a0a09acc460b825f8bdf7"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="GeneticAlgorithms_8py.html#a219a0bde9e1a0a09acc460b825f8bdf7">GeneticAlgorithms.pwd</a> (args)</td></tr>
<tr class="memdesc:a219a0bde9e1a0a09acc460b825f8bdf7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Password guessing.  <a href="GeneticAlgorithms_8py.html#a219a0bde9e1a0a09acc460b825f8bdf7">More...</a><br /></td></tr>
<tr class="separator:a219a0bde9e1a0a09acc460b825f8bdf7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a21c124df3b0885635ca2e7be8e28065d"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="GeneticAlgorithms_8py.html#a21c124df3b0885635ca2e7be8e28065d">GeneticAlgorithms.animate</a> (ga)</td></tr>
<tr class="memdesc:a21c124df3b0885635ca2e7be8e28065d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Plot animated progress of ga.  <a href="GeneticAlgorithms_8py.html#a21c124df3b0885635ca2e7be8e28065d">More...</a><br /></td></tr>
<tr class="separator:a21c124df3b0885635ca2e7be8e28065d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a79222700c1e03319bfeed67490f7d09a"><td class="memItemLeft" align="right" valign="top"><a id="a79222700c1e03319bfeed67490f7d09a"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="GeneticAlgorithms_8py.html#a79222700c1e03319bfeed67490f7d09a">GeneticAlgorithms.demo</a> ()</td></tr>
<tr class="memdesc:a79222700c1e03319bfeed67490f7d09a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Interactive demo for GeneticAlgorithms. <br /></td></tr>
<tr class="separator:a79222700c1e03319bfeed67490f7d09a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a67c3347f4b6d2b86c1233e6c2ddbae36"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="GeneticAlgorithms_8py.html#a67c3347f4b6d2b86c1233e6c2ddbae36">GeneticAlgorithms.main</a> ()</td></tr>
<tr class="memdesc:a67c3347f4b6d2b86c1233e6c2ddbae36"><td class="mdescLeft">&#160;</td><td class="mdescRight">Main Program - Unit Test and Demo.  <a href="GeneticAlgorithms_8py.html#a67c3347f4b6d2b86c1233e6c2ddbae36">More...</a><br /></td></tr>
<tr class="separator:a67c3347f4b6d2b86c1233e6c2ddbae36"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><dl class="section version"><dt>Version</dt><dd>1.1.0</dd></dl>
<dl class="section user"><dt>Purpose</dt><dd>Class to implement Genetic Algorithms using genes taking on values of settable alphabets (alleles) to form chromosomes which then are used to optimize parameters of an arbitrary objective function employing procedures from biologic evolution, i.e. "selective mating," "crossover," "mutation," "selection of the fittest," and "inversion."</dd></dl>
<dl class="section user"><dt>Comments</dt><dd>This is Python 3 code!</dd></dl>
<dl class="section user"><dt>Known Bugs</dt><dd>None</dd></dl>
<dl class="section author"><dt>Author</dt><dd>Ekkehard Blanz <a href="#" onclick="location.href='mai'+'lto:'+'Ekk'+'eh'+'ard'+'.B'+'lan'+'z@'+'gma'+'il'+'.co'+'m'; return false;">Ekkeh<span style="display: none;">.nosp@m.</span>ard.<span style="display: none;">.nosp@m.</span>Blanz<span style="display: none;">.nosp@m.</span>@gma<span style="display: none;">.nosp@m.</span>il.co<span style="display: none;">.nosp@m.</span>m</a> (C) 2019</dd></dl>
<dl class="section copyright"><dt>Copyright</dt><dd>See COPYING file that comes with this distribution </dd></dl>
</div><h2 class="groupheader">Function Documentation</h2>
<a id="file_a21c124df3b0885635ca2e7be8e28065d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#file_a21c124df3b0885635ca2e7be8e28065d">&#9670;&nbsp;</a></span>animate()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.animate </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>ga</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Plot animated progress of ga. </p>
<p>For regular optimization cases, this function plots the objective function once and then the individuals of each generation as red dots on it; for TSP cases, it plots the cities and the identified best path for each generation. </p>

</div>
</div>
<a id="file_a67c3347f4b6d2b86c1233e6c2ddbae36"></a>
<h2 class="memtitle"><span class="permalink"><a href="#file_a67c3347f4b6d2b86c1233e6c2ddbae36">&#9670;&nbsp;</a></span>main()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.main </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Main Program - Unit Test and Demo. </p>
<p>Unit test for GeneticAlgorithms class and demo for its use and function. </p>

</div>
</div>
<a id="file_aa9c763ce6c1837e80b6dc3abb2bbb2b5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#file_aa9c763ce6c1837e80b6dc3abb2bbb2b5">&#9670;&nbsp;</a></span>objfunc0()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.objfunc0 </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>args</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Objective function for demo from Goldberg's book (modified). </p>
<p>This function takes on a single float argument in [0 .. 1]. The original function exhibits one maximum at the upper end of the interval, which would be trivially maximized by this class as it always includes the chromosomes with all ones (and that with all zeros) in the starting population. This function is therefore slightly modified and exhibits the maximum at the center of the interval and a mirror image of itself in the upper half of the interval. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">*args</td><td>args[0] argument of objective function in [0 .. 1] </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>value of function [0 .. 1] </dd></dl>

</div>
</div>
<a id="file_ad568d0ddc86966d7809d304b52765e88"></a>
<h2 class="memtitle"><span class="permalink"><a href="#file_ad568d0ddc86966d7809d304b52765e88">&#9670;&nbsp;</a></span>objfunc1()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.objfunc1 </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>args</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Objective function 1 for demo. </p>
<p>This function takes on a single float argument in [0 .. 1]. It exhibits two maxima in this interval, one at x=0.15 with objective value 1, the other at x=0.7 with objective value 0.82. About 18.3% of the domain supports values above 0.82 and thus a population that can produce an individual with maximal fitness. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">*args</td><td>args[0] argument of objective function in [0 .. 1] </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>value of function [0 .. 1] </dd></dl>

</div>
</div>
<a id="file_a84f4a4e55c489adb038b679b81e0beae"></a>
<h2 class="memtitle"><span class="permalink"><a href="#file_a84f4a4e55c489adb038b679b81e0beae">&#9670;&nbsp;</a></span>objfunc2()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.objfunc2 </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>args</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Objective function 2 for demo. </p>
<p>This function takes on a single float argument in [0 .. 1]. It exhibits three maxima in this interval, one at x=0.15 with objective value 1, the other two both very close to x=0.8 yielding a very broad second maximum with objective value around 0.85. About 10.8% of the domain supports values above the second maximum and thus a population that can produce an individual with maximal fitness. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">*args</td><td>args[0] argument of objective function in [0 .. 1] </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>value of function [0 .. 1] </dd></dl>

</div>
</div>
<a id="file_ad4e5726cd3b24d9b11968b7a920c28c5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#file_ad4e5726cd3b24d9b11968b7a920c28c5">&#9670;&nbsp;</a></span>printHelp()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.printHelp </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print command line help text. </p>
<dl class="section return"><dt>Returns</dt><dd>0 </dd></dl>

</div>
</div>
<a id="file_a219a0bde9e1a0a09acc460b825f8bdf7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#file_a219a0bde9e1a0a09acc460b825f8bdf7">&#9670;&nbsp;</a></span>pwd()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.pwd </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>args</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Password guessing. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">args</td><td>tuple with string with password guess </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>percentage of correct characters at correct location [0 .. 1] </dd></dl>

</div>
</div>
<a id="file_a9b341db0c7dda40ecd30bc6491a0832f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#file_a9b341db0c7dda40ecd30bc6491a0832f">&#9670;&nbsp;</a></span>tsp()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.tsp </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>args</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Traveling Salesman Problem objective function. </p>
<p>The function returns the inverse of the distance a salesman would travel given the provided tour of cities, given by their indices. When the function is called for the first time (or when the parameter tsp.init is set to False), the function initializes a new set of city coordinates and inter-city distance matrix which is then kept constant through subsequent calls to tsp. If the number of cities is less than 12, the optimal path is computed and the returned value is then optDist / distance - otherwise, opDist is estimated and the result is then usually not bound by 1. The coordinates of the cities are made available via the parameter tsp.coordinates. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">*args</td><td>set of arguments (only the one element - array of coordinates) </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>inverse of total distance (in [0 .. 1] for small number of cities) </dd></dl>

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.13
</small></address>
</body>
</html>
