<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.2"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>GeneticAlgorithms: GeneticAlgorithms.GeneticAlgorithms Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">GeneticAlgorithms
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.2 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>GeneticAlgorithms</b></li><li class="navelem"><a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html">GeneticAlgorithms</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="class_genetic_algorithms_1_1_genetic_algorithms-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">GeneticAlgorithms.GeneticAlgorithms Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Genetic Algorithms (GA) optimizer for arbitrary objective functions.  
 <a href="class_genetic_algorithms_1_1_genetic_algorithms.html#details">More...</a></p>
<div class="dynheader">
Inheritance diagram for GeneticAlgorithms.GeneticAlgorithms:</div>
<div class="dyncontent">
 <div class="center">
  <img src="class_genetic_algorithms_1_1_genetic_algorithms.png" alt=""/>
 </div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a6ba3a6a8632cbfe20c5a35c7e01c7e0b"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#a6ba3a6a8632cbfe20c5a35c7e01c7e0b">__init__</a> (self, objectiveFunction, numberChromosomes, chromosomeLengths, populationSize, <a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#ae8991bbadaabcc69ff5a2cd5878caebb">decoder</a>=None, alleleAlphabet=None, <a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#a13fc2a169dd5849618f07b5d096401ba">pmx</a>=False, <a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#aacf531bd1dbf717add0804190ca7994f">pCrossover</a>=None, <a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#a8e00ccef261c502b07a984bfaec0bbfd">pMutation</a>=None, <a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#a9cd296e76e42b75df4208f3892f4f71b">pInversion</a>=None, <a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#a401fc77c68e49908abd8247c0fb1f6a2">populationGrowth</a>=1.0, <a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#a3b17ce20ca046d611b687ab99d688c4a">overpopulation</a>=1.3, chromosomeSets=1, <a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#a84cb79dc4970fa02b0e7553d18a4e8cc">fitnessScale</a>=0., monogamous=False, numberChildren=2, bestImmortal=True, <a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#acb21fb2bf7434de9a117e97667f857e1">floatSigma</a>=1.2, <a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#ab3ea578054d11005078798bd75d1e1f1">floatSigmaAdapt</a>=0.85, hook=None, threaded=False)</td></tr>
<tr class="memdesc:a6ba3a6a8632cbfe20c5a35c7e01c7e0b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constructor - check validity of parameters and set up first random population.  <a href="class_genetic_algorithms_1_1_genetic_algorithms.html#a6ba3a6a8632cbfe20c5a35c7e01c7e0b">More...</a><br /></td></tr>
<tr class="separator:a6ba3a6a8632cbfe20c5a35c7e01c7e0b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a48fdba4a535c324bb6499c51f2bf2ae7"><td class="memItemLeft" align="right" valign="top"><a id="a48fdba4a535c324bb6499c51f2bf2ae7" name="a48fdba4a535c324bb6499c51f2bf2ae7"></a>
def&#160;</td><td class="memItemRight" valign="bottom"><b>__str__</b> (self)</td></tr>
<tr class="memdesc:a48fdba4a535c324bb6499c51f2bf2ae7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assemble and return string with current internal parameters and current state of the population. <br /></td></tr>
<tr class="separator:a48fdba4a535c324bb6499c51f2bf2ae7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7b7757e0bff1f228ab22ec0ec64b73ca"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#a7b7757e0bff1f228ab22ec0ec64b73ca">run</a> (self, generations, maxfit=None)</td></tr>
<tr class="memdesc:a7b7757e0bff1f228ab22ec0ec64b73ca"><td class="mdescLeft">&#160;</td><td class="mdescRight">Run a defined number of generations or until maximum fitness is reached.  <a href="class_genetic_algorithms_1_1_genetic_algorithms.html#a7b7757e0bff1f228ab22ec0ec64b73ca">More...</a><br /></td></tr>
<tr class="separator:a7b7757e0bff1f228ab22ec0ec64b73ca"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a79de78fe612ed87ec24b7dfa36220240"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#a79de78fe612ed87ec24b7dfa36220240">objfunc</a> (self)</td></tr>
<tr class="memdesc:a79de78fe612ed87ec24b7dfa36220240"><td class="mdescLeft">&#160;</td><td class="mdescRight">Public read-only attribute to obtain the objective function.  <a href="class_genetic_algorithms_1_1_genetic_algorithms.html#a79de78fe612ed87ec24b7dfa36220240">More...</a><br /></td></tr>
<tr class="separator:a79de78fe612ed87ec24b7dfa36220240"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae8991bbadaabcc69ff5a2cd5878caebb"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#ae8991bbadaabcc69ff5a2cd5878caebb">decoder</a> (self)</td></tr>
<tr class="memdesc:ae8991bbadaabcc69ff5a2cd5878caebb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Public read-only attribute to obtain the decoder function.  <a href="class_genetic_algorithms_1_1_genetic_algorithms.html#ae8991bbadaabcc69ff5a2cd5878caebb">More...</a><br /></td></tr>
<tr class="separator:ae8991bbadaabcc69ff5a2cd5878caebb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a34dc116a93f77315f5275ff9c3033b2c"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#a34dc116a93f77315f5275ff9c3033b2c">statistics</a> (self)</td></tr>
<tr class="memdesc:a34dc116a93f77315f5275ff9c3033b2c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Public read-only attribute to obtain list of statistics dictionaries.  <a href="class_genetic_algorithms_1_1_genetic_algorithms.html#a34dc116a93f77315f5275ff9c3033b2c">More...</a><br /></td></tr>
<tr class="separator:a34dc116a93f77315f5275ff9c3033b2c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aca7186af90dbbf120c207f9cb7f65638"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#aca7186af90dbbf120c207f9cb7f65638">generation</a> (self)</td></tr>
<tr class="memdesc:aca7186af90dbbf120c207f9cb7f65638"><td class="mdescLeft">&#160;</td><td class="mdescRight">Public read-only attribute to obtain the current generation number with 0 being the starting generation.  <a href="class_genetic_algorithms_1_1_genetic_algorithms.html#aca7186af90dbbf120c207f9cb7f65638">More...</a><br /></td></tr>
<tr class="separator:aca7186af90dbbf120c207f9cb7f65638"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5e7f3b8bfc872e3431048d51ca66d40d"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#a5e7f3b8bfc872e3431048d51ca66d40d">population</a> (self)</td></tr>
<tr class="memdesc:a5e7f3b8bfc872e3431048d51ca66d40d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Public read-only attribute to obtain the current population of <a class="el" href="class_genetic_algorithms_1_1_genotype.html" title="Object representation of a genotype including genome and its properties and fitness of the correspond...">Genotype</a> instances.  <a href="class_genetic_algorithms_1_1_genetic_algorithms.html#a5e7f3b8bfc872e3431048d51ca66d40d">More...</a><br /></td></tr>
<tr class="separator:a5e7f3b8bfc872e3431048d51ca66d40d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a13fc2a169dd5849618f07b5d096401ba"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#a13fc2a169dd5849618f07b5d096401ba">pmx</a> (self)</td></tr>
<tr class="memdesc:a13fc2a169dd5849618f07b5d096401ba"><td class="mdescLeft">&#160;</td><td class="mdescRight">Public read-only attribute to obtain pmx flag.  <a href="class_genetic_algorithms_1_1_genetic_algorithms.html#a13fc2a169dd5849618f07b5d096401ba">More...</a><br /></td></tr>
<tr class="separator:a13fc2a169dd5849618f07b5d096401ba"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af9d83bbaffc7d533bfcf6e237fef2c96"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#af9d83bbaffc7d533bfcf6e237fef2c96">bestFit</a> (self)</td></tr>
<tr class="memdesc:af9d83bbaffc7d533bfcf6e237fef2c96"><td class="mdescLeft">&#160;</td><td class="mdescRight">Public read-only attribute to obtain decoded parameter and fitness of best performing individual in current population as a tuple.  <a href="class_genetic_algorithms_1_1_genetic_algorithms.html#af9d83bbaffc7d533bfcf6e237fef2c96">More...</a><br /></td></tr>
<tr class="separator:af9d83bbaffc7d533bfcf6e237fef2c96"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-static-methods" name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:ad234d826acb87117c24ce66aab956970"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#ad234d826acb87117c24ce66aab956970">genericDecoder</a> (genotype)</td></tr>
<tr class="memdesc:ad234d826acb87117c24ce66aab956970"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generic decoder - static member function to map a biallelic or triallelic genotype into a phenotype consisting of a numpy array of either real values all in the unit interval or a (list of) strings.  <a href="class_genetic_algorithms_1_1_genetic_algorithms.html#ad234d826acb87117c24ce66aab956970">More...</a><br /></td></tr>
<tr class="separator:ad234d826acb87117c24ce66aab956970"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-attribs" name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:aacf531bd1dbf717add0804190ca7994f"><td class="memItemLeft" align="right" valign="top"><a id="aacf531bd1dbf717add0804190ca7994f" name="aacf531bd1dbf717add0804190ca7994f"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>pCrossover</b></td></tr>
<tr class="memdesc:aacf531bd1dbf717add0804190ca7994f"><td class="mdescLeft">&#160;</td><td class="mdescRight">The probability of gene-crossover on a per chromosome gene basis. <br /></td></tr>
<tr class="separator:aacf531bd1dbf717add0804190ca7994f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8e00ccef261c502b07a984bfaec0bbfd"><td class="memItemLeft" align="right" valign="top"><a id="a8e00ccef261c502b07a984bfaec0bbfd" name="a8e00ccef261c502b07a984bfaec0bbfd"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>pMutation</b></td></tr>
<tr class="memdesc:a8e00ccef261c502b07a984bfaec0bbfd"><td class="mdescLeft">&#160;</td><td class="mdescRight">The probability for a gene to mutate on a per gene basis. <br /></td></tr>
<tr class="separator:a8e00ccef261c502b07a984bfaec0bbfd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9cd296e76e42b75df4208f3892f4f71b"><td class="memItemLeft" align="right" valign="top"><a id="a9cd296e76e42b75df4208f3892f4f71b" name="a9cd296e76e42b75df4208f3892f4f71b"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>pInversion</b></td></tr>
<tr class="memdesc:a9cd296e76e42b75df4208f3892f4f71b"><td class="mdescLeft">&#160;</td><td class="mdescRight">The probability for a sequence of genes to be inverted. <br /></td></tr>
<tr class="separator:a9cd296e76e42b75df4208f3892f4f71b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a401fc77c68e49908abd8247c0fb1f6a2"><td class="memItemLeft" align="right" valign="top"><a id="a401fc77c68e49908abd8247c0fb1f6a2" name="a401fc77c68e49908abd8247c0fb1f6a2"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>populationGrowth</b></td></tr>
<tr class="memdesc:a401fc77c68e49908abd8247c0fb1f6a2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Factor of net population growth per generation. <br /></td></tr>
<tr class="separator:a401fc77c68e49908abd8247c0fb1f6a2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3b17ce20ca046d611b687ab99d688c4a"><td class="memItemLeft" align="right" valign="top"><a id="a3b17ce20ca046d611b687ab99d688c4a" name="a3b17ce20ca046d611b687ab99d688c4a"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>overpopulation</b></td></tr>
<tr class="memdesc:a3b17ce20ca046d611b687ab99d688c4a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Over-population factor before selection of the fittest happens. <br /></td></tr>
<tr class="separator:a3b17ce20ca046d611b687ab99d688c4a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a84cb79dc4970fa02b0e7553d18a4e8cc"><td class="memItemLeft" align="right" valign="top"><a id="a84cb79dc4970fa02b0e7553d18a4e8cc" name="a84cb79dc4970fa02b0e7553d18a4e8cc"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>fitnessScale</b></td></tr>
<tr class="memdesc:a84cb79dc4970fa02b0e7553d18a4e8cc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Artificial spread of the fitness function (1.2 to 2.0) <br /></td></tr>
<tr class="separator:a84cb79dc4970fa02b0e7553d18a4e8cc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acb21fb2bf7434de9a117e97667f857e1"><td class="memItemLeft" align="right" valign="top"><a id="acb21fb2bf7434de9a117e97667f857e1" name="acb21fb2bf7434de9a117e97667f857e1"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>floatSigma</b></td></tr>
<tr class="memdesc:acb21fb2bf7434de9a117e97667f857e1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sigma for random numbers varying floating point genes. <br /></td></tr>
<tr class="separator:acb21fb2bf7434de9a117e97667f857e1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab3ea578054d11005078798bd75d1e1f1"><td class="memItemLeft" align="right" valign="top"><a id="ab3ea578054d11005078798bd75d1e1f1" name="ab3ea578054d11005078798bd75d1e1f1"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>floatSigmaAdapt</b></td></tr>
<tr class="memdesc:ab3ea578054d11005078798bd75d1e1f1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Modification factor for sigma floatSigma. <br /></td></tr>
<tr class="separator:ab3ea578054d11005078798bd75d1e1f1"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-static-attribs" name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:a48d4af0614144d849e1bae9a1aef3d65"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#a48d4af0614144d849e1bae9a1aef3d65">alphaAlphabet</a></td></tr>
<tr class="memdesc:a48d4af0614144d849e1bae9a1aef3d65"><td class="mdescLeft">&#160;</td><td class="mdescRight">alphabet (list) containing only ASCII letters and a blank  <a href="class_genetic_algorithms_1_1_genetic_algorithms.html#a48d4af0614144d849e1bae9a1aef3d65">More...</a><br /></td></tr>
<tr class="separator:a48d4af0614144d849e1bae9a1aef3d65"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1e798eb9aa007c6fc7eef82c0ef0cf80"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#a1e798eb9aa007c6fc7eef82c0ef0cf80">alnumAlphabet</a></td></tr>
<tr class="memdesc:a1e798eb9aa007c6fc7eef82c0ef0cf80"><td class="mdescLeft">&#160;</td><td class="mdescRight">alphabet (list) containing alphanumeric ASCII characters  <a href="class_genetic_algorithms_1_1_genetic_algorithms.html#a1e798eb9aa007c6fc7eef82c0ef0cf80">More...</a><br /></td></tr>
<tr class="separator:a1e798eb9aa007c6fc7eef82c0ef0cf80"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a19761456ebf58e64f95fd0a190b9956e"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html#a19761456ebf58e64f95fd0a190b9956e">characterAlphabet</a></td></tr>
<tr class="memdesc:a19761456ebf58e64f95fd0a190b9956e"><td class="mdescLeft">&#160;</td><td class="mdescRight">alphabet (list) containing characters on an American keyboard  <a href="class_genetic_algorithms_1_1_genetic_algorithms.html#a19761456ebf58e64f95fd0a190b9956e">More...</a><br /></td></tr>
<tr class="separator:a19761456ebf58e64f95fd0a190b9956e"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p >Genetic Algorithms (GA) optimizer for arbitrary objective functions. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="a6ba3a6a8632cbfe20c5a35c7e01c7e0b" name="a6ba3a6a8632cbfe20c5a35c7e01c7e0b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6ba3a6a8632cbfe20c5a35c7e01c7e0b">&#9670;&nbsp;</a></span>__init__()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.GeneticAlgorithms.__init__ </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>objectiveFunction</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>numberChromosomes</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chromosomeLengths</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>populationSize</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>decoder</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>alleleAlphabet</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pmx</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pCrossover</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pMutation</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pInversion</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>populationGrowth</em> = <code>1.0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>overpopulation</em> = <code>1.3</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>chromosomeSets</em> = <code>1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>fitnessScale</em> = <code>0.</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>monogamous</em> = <code>False</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>numberChildren</em> = <code>2</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>bestImmortal</em> = <code>True</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>floatSigma</em> = <code>1.2</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>floatSigmaAdapt</em> = <code>0.85</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>hook</em> = <code>None</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>threaded</em> = <code>False</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Constructor - check validity of parameters and set up first random population. </p>
<p >Default values are given in parameters. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">objectiveFunction</td><td>user-supplied Python function </td></tr>
    <tr><td class="paramname">numberChromosomes</td><td>number of chromosomes </td></tr>
    <tr><td class="paramname">chromosomeLengths</td><td>list of lengths of chromosomes - can be int if all chromosomes have the same length </td></tr>
    <tr><td class="paramname">populationSize</td><td>size of initial population </td></tr>
    <tr><td class="paramname">decoder</td><td>user-suppliedPython function (genericDecoder) </td></tr>
    <tr><td class="paramname">alleleAlphabet</td><td>alphabet list or type float from which to draw alleles ([0, 1] for haploid and [-1,0,1] for diploid chromosomes) </td></tr>
    <tr><td class="paramname">pmx</td><td>set True to enable partially matched crossover (False) </td></tr>
    <tr><td class="paramname">pCrossover</td><td>crossover probability (depending on method) </td></tr>
    <tr><td class="paramname">pMutation</td><td>mutation probability (depending on method) </td></tr>
    <tr><td class="paramname">pInversion</td><td>inversion probability </td></tr>
    <tr><td class="paramname">populationGrowth</td><td>growth factor between populations (1.0) </td></tr>
    <tr><td class="paramname">overpopulation</td><td>population overshoot before selection (1.2) </td></tr>
    <tr><td class="paramname">chromosomeSets</td><td>number of complete chromosome sets (1 - haploid) </td></tr>
    <tr><td class="paramname">fitnessScale</td><td>factor for fitness scaling (1.6 for regular chromosomes and None for float and pmx) </td></tr>
    <tr><td class="paramname">monogamous</td><td>monogamy parameter (False) </td></tr>
    <tr><td class="paramname">numberChildren</td><td>number of children per mating (2) </td></tr>
    <tr><td class="paramname">bestImmortal</td><td>set False to not have the best individual survive unchanged to the next generation (True) </td></tr>
    <tr><td class="paramname">floatSigma</td><td>standard deviation for float-type chromosome mutation (1.2) </td></tr>
    <tr><td class="paramname">floatSigmaAdapt</td><td>standard deviation adaptation for random mutation (0.85) </td></tr>
    <tr><td class="paramname">hook</td><td>user-supplied hook function called after every generation (None) </td></tr>
    <tr><td class="paramname">threaded</td><td>use threads when set to True (False) </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="af9d83bbaffc7d533bfcf6e237fef2c96" name="af9d83bbaffc7d533bfcf6e237fef2c96"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af9d83bbaffc7d533bfcf6e237fef2c96">&#9670;&nbsp;</a></span>bestFit()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.GeneticAlgorithms.bestFit </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Public read-only attribute to obtain decoded parameter and fitness of best performing individual in current population as a tuple. </p>
<p >This is NOT a method or member function! </p>

</div>
</div>
<a id="ae8991bbadaabcc69ff5a2cd5878caebb" name="ae8991bbadaabcc69ff5a2cd5878caebb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae8991bbadaabcc69ff5a2cd5878caebb">&#9670;&nbsp;</a></span>decoder()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.GeneticAlgorithms.decoder </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Public read-only attribute to obtain the decoder function. </p>
<p >This is NOT a method or member function! </p>

</div>
</div>
<a id="aca7186af90dbbf120c207f9cb7f65638" name="aca7186af90dbbf120c207f9cb7f65638"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aca7186af90dbbf120c207f9cb7f65638">&#9670;&nbsp;</a></span>generation()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.GeneticAlgorithms.generation </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Public read-only attribute to obtain the current generation number with 0 being the starting generation. </p>
<p >This is NOT a method or member function! </p>

</div>
</div>
<a id="ad234d826acb87117c24ce66aab956970" name="ad234d826acb87117c24ce66aab956970"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad234d826acb87117c24ce66aab956970">&#9670;&nbsp;</a></span>genericDecoder()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.GeneticAlgorithms.genericDecoder </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>genotype</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Generic decoder - static member function to map a biallelic or triallelic genotype into a phenotype consisting of a numpy array of either real values all in the unit interval or a (list of) strings. </p>
<p >For numeric biallelic or triallelic chromosomes, all chromosomes in the genotype will be mapped into a single real value in [0 .. 1] each, treating each gene as a bit of an integer and then mapping the integer onto [0 .. 1]. In the case of floating point genes or if the partially matched crossover (pmx) flag is set, they are returned directly as an array, one array for each chromosome. If there was only one biallelic or triallelic chromosome, the return value will not be an array but rather a single real value in [0 .. 1] instead. The same is true for a single valued chromosome. For chromosomes whose alleles are characters, the return value is either a single string or a list of string if there was more than one chromosome. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">genotype</td><td>instance of class <a class="el" href="class_genetic_algorithms_1_1_genotype.html" title="Object representation of a genotype including genome and its properties and fitness of the correspond...">Genotype</a> </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>tuple consisting of either a numpy array, a single real value, a string, or a list of strings </dd></dl>

</div>
</div>
<a id="a79de78fe612ed87ec24b7dfa36220240" name="a79de78fe612ed87ec24b7dfa36220240"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a79de78fe612ed87ec24b7dfa36220240">&#9670;&nbsp;</a></span>objfunc()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.GeneticAlgorithms.objfunc </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Public read-only attribute to obtain the objective function. </p>
<p >This is NOT a method or member function! </p>

</div>
</div>
<a id="a13fc2a169dd5849618f07b5d096401ba" name="a13fc2a169dd5849618f07b5d096401ba"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a13fc2a169dd5849618f07b5d096401ba">&#9670;&nbsp;</a></span>pmx()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.GeneticAlgorithms.pmx </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Public read-only attribute to obtain pmx flag. </p>
<p >This is NOT a method or member function! </p>

</div>
</div>
<a id="a5e7f3b8bfc872e3431048d51ca66d40d" name="a5e7f3b8bfc872e3431048d51ca66d40d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5e7f3b8bfc872e3431048d51ca66d40d">&#9670;&nbsp;</a></span>population()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.GeneticAlgorithms.population </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Public read-only attribute to obtain the current population of <a class="el" href="class_genetic_algorithms_1_1_genotype.html" title="Object representation of a genotype including genome and its properties and fitness of the correspond...">Genotype</a> instances. </p>
<p >This is NOT a method or member function! </p>

</div>
</div>
<a id="a7b7757e0bff1f228ab22ec0ec64b73ca" name="a7b7757e0bff1f228ab22ec0ec64b73ca"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7b7757e0bff1f228ab22ec0ec64b73ca">&#9670;&nbsp;</a></span>run()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.GeneticAlgorithms.run </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>generations</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>maxfit</em> = <code>None</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Run a defined number of generations or until maximum fitness is reached. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">generations</td><td>number of generations to produce </td></tr>
    <tr><td class="paramname">maxfit</td><td>abortion criterion when fitness reached maximum (default is None) </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a34dc116a93f77315f5275ff9c3033b2c" name="a34dc116a93f77315f5275ff9c3033b2c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a34dc116a93f77315f5275ff9c3033b2c">&#9670;&nbsp;</a></span>statistics()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def GeneticAlgorithms.GeneticAlgorithms.statistics </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>self</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Public read-only attribute to obtain list of statistics dictionaries. </p>
<p >The keys into these dictionaries are: "mean", "variance", "min", "max", "crossovers", "mutations", and "divorcerate". Lists of individual statistics key can be obtained via [s[key] for s in ga.statistics] where ga is a <a class="el" href="class_genetic_algorithms_1_1_genetic_algorithms.html" title="Genetic Algorithms (GA) optimizer for arbitrary objective functions.">GeneticAlgorithms</a> object.</p>
<p >It is worth noting that these statistics are computed BEFORE the selection of the fittest took place (as they are needed to select them).</p>
<p >This is NOT a method or member function! </p>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a id="a1e798eb9aa007c6fc7eef82c0ef0cf80" name="a1e798eb9aa007c6fc7eef82c0ef0cf80"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1e798eb9aa007c6fc7eef82c0ef0cf80">&#9670;&nbsp;</a></span>alnumAlphabet</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">GeneticAlgorithms.GeneticAlgorithms.alnumAlphabet</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line">=  copy.copy(</div>
<div class="line">        alphaAlphabet )</div>
</div><!-- fragment -->
<p>alphabet (list) containing alphanumeric ASCII characters </p>

</div>
</div>
<a id="a48d4af0614144d849e1bae9a1aef3d65" name="a48d4af0614144d849e1bae9a1aef3d65"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a48d4af0614144d849e1bae9a1aef3d65">&#9670;&nbsp;</a></span>alphaAlphabet</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">GeneticAlgorithms.GeneticAlgorithms.alphaAlphabet</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line">=  list(</div>
<div class="line">        <span class="stringliteral">&quot; &quot;</span> </div>
<div class="line">        <span class="stringliteral">&quot;ABCDEFGHIJKLMNOPQRSTUVWXYZ&quot;</span> </div>
<div class="line">        <span class="stringliteral">&quot;abcdefghijklmnopqrstuvwxyz&quot;</span> )</div>
</div><!-- fragment -->
<p>alphabet (list) containing only ASCII letters and a blank </p>

</div>
</div>
<a id="a19761456ebf58e64f95fd0a190b9956e" name="a19761456ebf58e64f95fd0a190b9956e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a19761456ebf58e64f95fd0a190b9956e">&#9670;&nbsp;</a></span>characterAlphabet</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">GeneticAlgorithms.GeneticAlgorithms.characterAlphabet</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line">=  copy.copy(</div>
<div class="line">        alnumAlphabet )</div>
</div><!-- fragment -->
<p>alphabet (list) containing characters on an American keyboard </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li><a class="el" href="_genetic_algorithms_8py.html">GeneticAlgorithms.py</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.2
</small></address>
</body>
</html>
